org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/folding/DefaultFoldingRangeProvider.java

```diff
@@ -0,0 +1,197 @@
+						EObject eObject = allContents.next();
+						if (isHandled(eObject)) {
+							acceptObjectFolding(eObject, foldingRangeAcceptor);
+						}
+						if (!shouldProcessContent(eObject)) {
+							allContents.prune();
+						}
+					}
+				}
+			}
+		}
+	}
+
+	protected void acceptObjectFolding(EObject eObject, IFoldingRangeAcceptor foldingRangeAcceptor) {
+		ITextRegion region = locationInFileProvider.getFullTextRegion(eObject);
+		if (region != null) {
+			INode eObjectNode = NodeModelUtils.getNode(eObject);
+			ITextRegion significant = buildSignificantRegion(locationInFileProvider.getSignificantTextRegion(eObject),
+					eObjectNode);
+			foldingRangeAcceptor.accept(region.getOffset(), region.getLength(), null, false, significant);
+		}
+	}
+	
+	protected ITextRegion buildSignificantRegion(ITextRegion significantRegion, INode node) {
+
+		int offset = significantRegion.getOffset();
+		int endOffset = significantRegion.getOffset() + significantRegion.getLength();
+
+		int startLine;
+		int endLine;
+		if (significantRegion instanceof ITextRegionWithLineInformation) {
+			ITextRegionWithLineInformation lineInfoRegion = (ITextRegionWithLineInformation) significantRegion;
+			startLine = lineInfoRegion.getLineNumber();
+			endLine = lineInfoRegion.getEndLineNumber();
+		} else {
+			startLine = NodeModelUtils.getLineAndColumn(node, offset).getLine();
+			endLine = NodeModelUtils.getLineAndColumn(node, endOffset).getLine();
+		}
+
+		if (startLine != endLine) {
+			for (int index = offset; index < endOffset; index++) {
+				LineAndColumn lineInfo = NodeModelUtils.getLineAndColumn(node, index);
+				if (lineInfo.getLine() != startLine) {
+					return new TextRegion(offset, index - offset - 1);
+				}
+			}
+		}
+		return significantRegion;
+	}
+
+	protected void computeCommentFolding(XtextResource xtextDocument, IFoldingRangeAcceptor foldingRangeAcceptor) {
+		IParseResult parseResult = xtextDocument.getParseResult();
+		if (parseResult != null) {
+			EObject rootASTElement = parseResult.getRootASTElement();
+			acceptCommentNodes(rootASTElement, foldingRangeAcceptor);
+		}
+	}
+
+	protected void acceptCommentNodes(EObject eObject, IFoldingRangeAcceptor foldingRangeAcceptor) {
+		ICompositeNode node = NodeModelUtils.getNode(eObject);
+		if (node != null) {
+			for (INode leafNode : node.getAsTreeIterable()) {
+				if (leafNode.getGrammarElement() instanceof TerminalRule
+						&& ruleName.equalsIgnoreCase(((TerminalRule) leafNode.getGrammarElement()).getName())) {
+					acceptCommentFolding(leafNode, foldingRangeAcceptor);
+				}
+			}
+		}
+	}
+
+	protected void acceptCommentFolding(INode commentNode, IFoldingRangeAcceptor foldingRangeAcceptor) {
+		int offset = commentNode.getOffset();
+		int length = commentNode.getLength();
+		Matcher matcher = getTextPatternInComment().matcher(commentNode.getText());
+		if (matcher.find()) {
+			TextRegion significant = new TextRegion(offset + matcher.start(), 0);
```

Comment:
I'm not sure the significant part of a comment is the empty region. Shouldn't this be the first line?

Reply:
I just took the original code from [here](https://github.com/eclipse/xtext-eclipse/blob/aa0272eb77c5476e2a245812c5362628bc10d2d1/org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/folding/DefaultFoldingRegionProvider.java#L183-L194).

Reply:
Well, this means there is no text in the comment. The significant region for folding is probably used to show some text upon hovering, if there's no text in the comment there's not much to be shown on hover IMO so creating an empty significant region might be ok.